src/harness/simple_crl_server.py

```diff
@@ -0,0 +1,320 @@
+        The CRL server runs on this URL and responds only to this URL with the CRL file.
+      crl_file_path: Absolute path to the CRL file that will be served by this CRL server.
+    """
+
+    super(SimpleCrlServer, self).__init__()
+    url_components = urlparse(crl_url)
+    self.host_name = url_components.hostname
+    self.port = url_components.port if url_components.port is not None else \
+      DEFAULT_CRL_SERVER_PORT
+    self.crl_url_path = url_components.path
+    self.crl_file_path = crl_file_path
+    self.setDaemon(True)
+
+  def run(self):
+    """Starts the HTTPServer as background thread.
+
+    The run() method is an overridden method from thread class and it gets invoked
+    when the thread is started using thread.start().
+    """
+
+    # Setting the parameters for handler class.
+    logging.info("Started CRL Server at %s", self.host_name + ":" + str(self.port))
+    self.server = CrlHttpServer(self.crl_url_path, self.crl_file_path,
+                                (self.host_name, self.port),
+                                CrlServerHttpHandler)
+    self.server.serve_forever()
+
+  def stopServer(self):
+    """This method is used to stop HTTPServer Socket"""
+
+    self.server.shutdown()
+    logging.info('Stopped CRL Server at %s', self.host_name + ":" + str(self.port))
+
+class CrlHttpServer(HTTPServer):
+  """The class CrlHttpServer overrides built-in HTTPServer.
+
+  It takes the parameter base_path used to serve the files.
+  This is needed to have CRL Server to serve files from different directories.
+  """
+
+  def __init__(self, crl_url_path, crl_file_path, server_address, RequestHandlerClass):
+    self.crl_url_path = crl_url_path
+    self.crl_file_path = crl_file_path
+    HTTPServer.__init__(self, server_address, RequestHandlerClass)
+
+class CrlServerHttpHandler(BaseHTTPRequestHandler):
+  """This class implements the HTTP handlers for the CRL Server.
+
+  CrlServerHttpHandler class inherits with BaseHTTPRequestHandler
+  to serve HTTP Response.
+  """
+
+  def do_GET(self):
+    """Handles Pull/GET Request and returns path of the request to callback method."""
+
+    parsed_path = urlparse(self.path)
+    self.log_message('Received GET Request:{}'.format(parsed_path.path))
+
+    # Handles the GET Requests.
+    # If the GET request path does not match the CRL URL path then respond with 404 response.
+    if parsed_path.path != self.server.crl_url_path:
+      self.log_message('Requested url :{0} does not match with CRL Server URL '
+                       'path:{1}'.format(parsed_path.path, self.server.crl_url_path))
+      self.send_response(404)
+      return
+
+    # Load the ca.crl and write content in HTTP response stream with 200 OK response.
+    # If any exception occurs while reading CRL chain and handle error and respond
+    # with 404 'Not Found' response.
+    try:
+      with open(os.path.join(self.server.crl_file_path)) as file_handle:
+        self.wfile.write(file_handle.read())
```

Comment:
When I try to fetch http://localhost:9007/ca.crl from Chrome, it fails with ERR_INVALID_HTTP_RESPONSE, because the data arrives before the headers.  I recommend this pattern:

    try:
      with open(os.path.join(self.server.crl_file_path)) as file_handle:
        file_data = file_handle.read()
    except IOError:
      ...
    else:
      ...
      self.end_headers()
      self.wfile.write(file_data)

(This also fixes the edge case where a `write()` error could trigger a 404 response.)

Reply:
Thanks   for catching this. We tested using 'wget' and it was working. We see that it   is not working with chrome. The code has been fixed as suggested.